### ISSUE TITLE
AEOSVDC increasing CPU

### ISSUE THREAD

**Issue by FreddieMatherSmartDCSIT** [xtuml/munin#156]
Behaviour of the CPU measurements has been noted throughout longer run endurance tests. The CPU usage of AEOSVDC containers increases with a trend of a steady in a linear increase even when it seems that AEOSVDC is able to process the the sent events at the required rate. It has been noted that event response times (time from sending to fully processed by the PV) also show this linear trend. This has been noted in runs with both 4 AEOSVDC (8 hour test and 24 hour test)  and 8 AEOSVDC (24 hour test) containers. When the CPU maxes out for all containers the response times show a highly non-linear increase.

### 8 hour run (4 AEOSVDC)

The 8 hour run had the following container numbers and events sent/s

- 500 events per second sent for 8 hours
- 1 AEReception container
- 4 AEOSVDC containers
- PV version 1.1.1

CPU linearly increasing as below and maxing out

![image](https://github.com/xtuml/munin/assets/124710637/ecb03822-a061-422d-9277-e30fe7977227)

Response times (aggregated mean values in one second bins) linearly increasing until CPU maxes out and then starting to non-linearly increase just before end of test

![image](https://github.com/xtuml/munin/assets/124710637/18cc41e6-42a5-46ad-9026-0db681e9d62d)

### 24 hour (4 AEO SVDC)

The 24 hour run had the following container numbers and events sent/s

- 500 events per second sent for 24 hours
- 1 AEReception container
- 4 AEOSVDC containers
- PV version 1.1.1

CPU linearly increasing as below and maxing out after approximately 8 hours

![image](https://github.com/xtuml/munin/assets/124710637/d6bedf0c-5fb8-4bd4-997c-23c66d9c8e19)

Response times (aggregated mean values in one second bins) linearly increasing until CPU maxes out and then non-linearly increase before no more events are processed

![image](https://github.com/xtuml/munin/assets/124710637/29c3f043-0f9a-4db3-a658-941d9a405e9a)

### 24 hour (8 AEO SVDC)

The 24 hour run had the following container numbers and events sent/s

- 500 events per second sent for 24 hours
- 1 AEReception container
- 8 AEOSVDC containers
- PV version 1.1.1

CPU linearly increasing at a slower rate than with 4 AEOSVDC containers. Doesn't max out but shows the behaviour still exists even. The expected trend would likely max out the cpu over another 1-2 days

![image](https://github.com/xtuml/munin/assets/124710637/9afdcab7-7870-48c0-9427-2bba3a5c7e8f)

Response times (aggregated mean values in one second bins) linearly increasing over the test

**Comment by cortlandstarrett** [comment]
We identified a memory leak.  We were not deleting instances of the PVjob and InstrumentationEvents.
Identified on 26 October.
Will be fixed in the next tag.

**Comment by cortlandstarrett** [comment]
<img width="474" alt="memory" src="https://github.com/xtuml/munin/assets/1993198/724187da-32e6-4650-99a0-746f01912273">

**Comment by cortlandstarrett** [comment]
a 2 million event run began at 15:46 and leveled out pretty quickly
This was running 800 events/second.

**Comment by FreddieMatherSmartDCSIT** [comment]
We see that kind of memory profile in the shorter time frames (note we send a linearly increasing profile of events at the start which is why you see the slow ramp up)

![image](https://github.com/xtuml/munin/assets/124710637/53ff9789-ab58-41b8-b615-7a469fe6dae9)

but when the cpu maxes out thats when events back up and the memory starts to spike

![image](https://github.com/xtuml/munin/assets/124710637/a13b9022-cb4b-49b7-9070-c3a0eb2852a2)

**Comment by cortlandstarrett** [comment]
Yes, understood.  The memory leak we found caused (causes) exactly this behaviour.  We were not deleting an element in a list that was repeatedly searched.  As the list got longer, the CPU time to search the list got longer.  We did not run out of memory before we ran out of CPU cycles which caused multiple jobs to be in process simultaneously  until we folded.

I will be running a 24 hour test soon to discover if there are any more resource issues hiding.
